File: apps/server/src/routes/device.ts
import { Router, Request, Response, NextFunction } from "express";
import { validate } from "../middleware/validate";
import { deviceAuthMiddleware } from "../middleware/deviceAuth";
import * as deviceService from "../services/deviceService";
import { z } from "zod";

const router = Router();

const asyncHandler =
  (fn: (req: Request, res: Response, next: NextFunction) => Promise<void>) =>
  (req: Request, res: Response, next: NextFunction) =>
    fn(req, res, next).catch(next);

// ===== 设备注册（无需签名，首次启动） =====
// 精简：只接收 device_id + platform

const registerDeviceSchema = z.object({
  device_id: z.string().regex(/^[0-9a-fA-F]{64}$/, "device_id must be 64-char hex (Ed25519 public key)"),
  platform: z.enum(["ios", "android", "web"], { required_error: "platform is required" }),
});

router.post(
  "/",
  validate(registerDeviceSchema),
  asyncHandler(async (req: Request, res: Response) => {
    const result = await deviceService.registerDevice(req.body);
    // 新设备返回 201，已存在设备返回 200（幂等）
    const status = result.created_at.getTime() === result.updated_at.getTime() ? 201 : 200;
    res.status(status).json(result);
  })
);

// ===== 获取当前设备信息（需要签名） =====

router.get(
  "/me",
  deviceAuthMiddleware,
  asyncHandler(async (req: Request, res: Response) => {
    const deviceId = req.headers["x-device-id"] as string;
    const result = await deviceService.getDevice(deviceId);
    res.json(result);
  })
);

// ===== 设备订阅钱包（需要签名） =====

const subscribeWalletSchema = z.object({
  wallet_id: z.string().min(1, "wallet_id is required"),
  chain: z.string().max(32).optional(),
  address_id: z.string().max(36).optional(),
});

router.post(
  "/wallets",
  deviceAuthMiddleware,
  validate(subscribeWalletSchema),
  asyncHandler(async (req: Request, res: Response) => {
    const deviceId = req.headers["x-device-id"] as string;
    const result = await deviceService.subscribeWallet(
      deviceId,
      req.body.wallet_id,
      req.body.chain,
      req.body.address_id
    );
    res.status(201).json(result);
  })
);

// ===== 设备取消订阅钱包（需要签名） =====

router.delete(
  "/wallets/:wallet_id",
  deviceAuthMiddleware,
  asyncHandler(async (req: Request, res: Response) => {
    const deviceId = req.headers["x-device-id"] as string;
    await deviceService.unsubscribeWallet(deviceId, req.params.wallet_id as string);
    res.status(204).send();
  })
);

// ===== 获取设备订阅的所有钱包（需要签名） =====

router.get(
  "/wallets",
  deviceAuthMiddleware,
  asyncHandler(async (req: Request, res: Response) => {
    const deviceId = req.headers["x-device-id"] as string;
    const result = await deviceService.getDeviceWallets(deviceId);
    res.json({ wallets: result });
  })
);

export default router;

File: apps/server/src/routes/device.rs
//! 设备路由 — /api/v1/devices (1 个接口: 注册设备)

use crate::errors::AppError;
use crate::middleware::AppState;
use crate::models::Device;
use crate::services::device_service;
use axum::{extract::State, routing::post, Json, Router};
use rbdc::DateTime;
use serde::{Deserialize, Serialize};

pub fn public_router() -> Router<AppState> {
    Router::new().route("/devices", post(register_device))
}

#[derive(Debug, Deserialize)]
struct RegisterDeviceRequest {
    device_id: String,
    platform: Option<String>,
}

#[derive(Debug, Serialize)]
struct DeviceResponse {
    id: String,
    platform: String,
    created_at: Option<DateTime>,
    updated_at: Option<DateTime>,
}

impl From<Device> for DeviceResponse {
    fn from(d: Device) -> Self {
        Self {
            id: d.id,
            platform: d.platform,
            created_at: d.created_at,
            updated_at: d.updated_at,
        }
    }
}

async fn register_device(
    State(state): State<AppState>,
    Json(body): Json<RegisterDeviceRequest>,
) -> Result<(axum::http::StatusCode, Json<DeviceResponse>), AppError> {
    let (device, created) = device_service::register_device(
        state.db.clone(),
        &body.device_id,
        body.platform.as_deref().unwrap_or("ios"),
    )
    .await?;
    let status = if created {
        axum::http::StatusCode::CREATED
    } else {
        axum::http::StatusCode::OK
    };
    Ok((status, Json(DeviceResponse::from(device))))
}
